js/event.ts

```diff
@@ -0,0 +1,242 @@
+
+    if (this.path.length === 0) {
+      return composedPath;
+    }
+
+    composedPath.push(this.currentTarget);
+
+    let currentTargetIndex = 0;
+    let currentTargetHiddenSubtreeLevel = 0;
+
+    for (let index = this.path.length - 1; index >= 0; index--) {
+      const { item, rootOfClosedTree, slotInClosedTree } = this.path[index];
+
+      if (rootOfClosedTree) {
+        currentTargetHiddenSubtreeLevel++;
+      }
+
+      if (item === this.currentTarget) {
+        currentTargetIndex = index;
+        break;
+      }
+
+      if (slotInClosedTree) {
+        currentTargetHiddenSubtreeLevel--;
+      }
+    }
+
+    let currentHiddenLevel = currentTargetHiddenSubtreeLevel;
+    let maxHiddenLevel = currentTargetHiddenSubtreeLevel;
+
+    for (let i = currentTargetIndex - 1; i >= 0; i--) {
+      const { item, rootOfClosedTree, slotInClosedTree } = this.path[i];
+
+      if (rootOfClosedTree) {
+        currentHiddenLevel++;
+      }
+
+      if (currentHiddenLevel <= maxHiddenLevel) {
+        composedPath.unshift(item);
+      }
+
+      if (slotInClosedTree) {
+        currentHiddenLevel--;
+
+        if (currentHiddenLevel < maxHiddenLevel) {
+          maxHiddenLevel = currentHiddenLevel;
+        }
+      }
+    }
+
+    currentHiddenLevel = currentTargetHiddenSubtreeLevel;
+    maxHiddenLevel = currentTargetHiddenSubtreeLevel;
+
+    for (let index = currentTargetIndex + 1; index < this.path.length; index++) {
+      const { item, rootOfClosedTree, slotInClosedTree } = this.path[index];
+
+      if (slotInClosedTree) {
+        currentHiddenLevel++;
+      }
+
+      if (currentHiddenLevel <= maxHiddenLevel) {
+        composedPath.push(item);
+      }
+
+      if (rootOfClosedTree) {
+        currentHiddenLevel--;
+
+        if (currentHiddenLevel < maxHiddenLevel) {
+          maxHiddenLevel = currentHiddenLevel;
+        }
+      }
+    }
+
+    return composedPath;
+  }
+
+  /** Cancels the event (if it is cancelable).
+   * See https://dom.spec.whatwg.org/#set-the-canceled-flag
+   *
+   *      event.preventDefault();
+   */
+  preventDefault(): void {
+    if (this.cancelable && !this.inPassiveListenerFlag) {
+      this.canceledFlag = true;
+    }
+  }
+
+  /** Stops the propagation of events further along in the DOM.
+   *
+   *      event.stopPropagation();
+   */
+  stopPropagation(): void {
+    this.stopPropagationFlag = true;
+  }
+
+  /** For this particular event, no other listener will be called.
+   * Neither those attached on the same element, nor those attached
+   * on elements which will be traversed later (in capture phase,
+   * for instance).
+   *
+   *      event.stopImmediatePropagation();
+   */
+  stopImmediatePropagation(): void {
+    this.stopPropagationFlag = true;
+    this.stopImmediatePropagationFlag = true;
+  }
+}
+
+// Built-in objects providing `get` methods for our interceptable JavaScript operations
+Reflect.defineProperty(Event.prototype, 'bubbles', { enumerable: true });
```

Comment:
I know what this does, but why are we doing this?

Reply:
I'm following the [IDL level 1 standard](https://www.w3.org/TR/WebIDL-1/#es-attributes) for attributes, though I swapped `Reflect` for `Object` since these are just properties.

Reply:
When you define the getter in the prototype, it is automatically enumerable.  So I am not sure what value this is supplying again.  If you are trying to get `configurable` to be false, I really don't think there is much value in that, but that only applies with the property being `[Unforgetable]`, but I would like to see a use case where this actually matters, or where this breaks API capability.

Reply:
Properties created with class syntax, including accessors, default to `enumerable: false`, so explicitly setting enumerability of prototype accessors to true after declaration is correct for Web IDL classes (unless TypeScript classes vary from JS in this regard?)

As for whether it matters — I think it would be very rare. But if it can be observed, it can be relied on, and even if surprising or unusual, it tends to be true that eventually somebody probably will, at least out on the “open web.” That’s the reason this stuff is described down to the last nut and bolt, anyway.

On the other hand, the necessity of conformance seems like a broad project goals question. Is the goal to produce a conformant implementation, or an implementation which is similar enough for some set of anticipated use cases? If the latter, maybe the rules for which facets of the APIs are deemed not worth implementing ought to be recorded somewhere for reference. Making case-by-case judgment calls about where to diverge from the spec seems like it might not play out great.

Reply:
Since members is still a proposal, TypeScript's non-accessor properties are non-compliant with the proposal.  There is an issue tracking it, but they are enumerable on the prototype.  Because accessors are part of the existing class standard, TypeScript just passes them through and yes, you are right they are not enumerable.

Given that they IDL specifically states they should be enumerable accessors, it makes sense to do it here and we should check other Web APIs.  Personally don't care about the `Reflect.defineProperty()` and Object.defineProperty()` but when you are ignoring the return result, there is really no benefit.  I think a more clear comment of what we are doing and why would help too.